[feat] initial draft of delegate server

MangoIV opened the pull request:
- add Delegate Server and types
- rewrite the Main to work with a proper server
- some changes to handle serialisation

MangoIV committed:
[feat] initial draft of delegate server
- add Delegate Server and types
- rewrite the Main to work with a proper server
- some changes to handle serialisation

uhbif19 commented on delegate-app/HydraAuction/Delegate/Server.hs:
```diff
@@ -0,0 +1,119 @@
+
+type DelegateTracer = TracerT DelegateServerLog
+
+class MonadTracer t m | m -> t where
```
Please separate transformer/monad stuff to separate file.

Also I do not understand what is need for this. Like consumer is just IO thread, it could just log into Tracer constant.

MangoIV replied:
yes, I just haven't done it yet, thank you

uhbif19 commented on delegate-app/HydraAuction/Delegate/Server.hs:
```diff
@@ -0,0 +1,119 @@
+    DelegateError err -> "Delegate error" <> line <> pretty err
+
+-- FIXME: needs actual error
+data DelegateError = FrontendNoParse
```
What would we do with error datatype?

Client should get error by async response anyway. So I guess this should be fit in regular `delegateStep` logic.

MangoIV replied:
well those are the errors that are emitted by the thing that calls `execDelegateRunner`, so we cannot emit them from inside the monad, e.g. this specific appears if we cannot parse the JSON that the frontend inputs

uhbif19 replied:
Yes. But still I not see need for Either. This code should just return some answer like `WrongInput` to client and add nothing to inputs queue. That is that Hydra does in such case AFAIU.

MangoIV replied:
Well yeah, this doesn't get enqueued anywhere

MangoIV replied:
Reminder to self to respond with this error if there occurs a failure in encoding

uhbif19 commented on delegate-app/HydraAuction/Delegate/Server.hs:
```diff
@@ -0,0 +1,119 @@
+
+instance Pretty DelegateServerLog where
+  pretty = \case
+    Started port -> "Started Server at Port" <> pretty (show port)
```
All other logs just use show so this may look strange in CLI.

MangoIV replied:
this is what IOG uses largely, I think it's quite nice because we can just in a modular way structure our logs and adjust the error messages whenever we want to. I don't see why it would look strange in CLI, it would just be formatted a bit differently, I think it's nicer than adhoc strings in the code.

MangoIV replied:
this also serves as documentation

uhbif19 replied:
> this is what IOG uses largely, I think it's quite nice because we can just in a modular way structure our logs and adjust the error messages whenever we want to. ..., I think it's nicer than adhoc strings in the code.

Of course, I completely agree.

> I don't see why it would look strange in CLI, it would just be formatted a bit differently, 

Because different formatting stands out when reading logs. Anyway this does not really matter, we could add some prefix or something else if that would be a problem.

uhbif19 commented on delegate-app/HydraAuction/Delegate/Server.hs:
```diff
@@ -0,0 +1,119 @@
+  }
+
+data DelegateServerLog
+  = Started PortNumber
```
Cool.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,133 @@
+  port <- lookupEnv "port"
+
+  -- FIXME: do we need this?
+  let host :: IP = "127.0.0.1"
```
I think while style is better cuz it makes clear that no monadic side-effects are happening.

uhbif19 replied:
*`where` style

nazrhom replied:
We can parse these as args (we will need it also for hydra node later). We can use a similar style as we use in the CLI

MangoIV replied:
> We can parse these as args (we will need it also for hydra node later). We can use a similar style as we use in the CLI

Is this in scope of this PR?

uhbif19 replied:
I guess we can separate this to follow-up issue to iterate quickly.

nazrhom replied:
I opened #243 to track this

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,133 @@
+  -- state? probably have to remodel this
+  execDelegateRunnerT $ flip evalStateT True $ mkRunner tick connection
+
+-- FIXME: remove concrete monads and make this tagless final (needs MonadDelegate)
```
Why remove concrete monads? This should just be called from main, no?

MangoIV replied:
I think this was mainly because it was easier to read the tagless final form than having concrete monads in this case. also we don't lose anything (perhaps performance but I don't even think that's the case here)

uhbif19 replied:
>  it was easier to read the tagless final form than having concrete monads

Why? `MonadDelegate` does not seem to have any laws or any logic at all.

> also we don't lose anything

Introduction of new interface adds code and necessity to read it.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,133 @@
+  TracerT DelegateServerLog IO ()
+runDelegateServer conf = do
+  trace (Started $ dlgt'port conf)
+  let fallback :: Application
```
What does this code do?

uhbif19 replied:
Like it is some HTTP-specific stuff. Hydra just uses `runServer` and that seems enough.

Hydra code is probably over complicated over what we need.

MangoIV replied:
in the docs it explicitly said you should not use `runServer`, that's why I didn't, this basically offers a fallback if the frontend tries to communicate via HTTP instead of websockets

MangoIV replied:
https://hackage.haskell.org/package/websockets-0.12.7.3/docs/Network-WebSockets.html#v:runServer

uhbif19 replied:
> this basically offers a fallback if the frontend tries to communicate via HTTP instead of websockets

But it should not do that. Our frontend is not browser.

> Glue for using this package with real servers is provided by:

I think they are talking about integration with HTTP server which we do not do anyway.

Their points: about DDoS is not relevant for our scope and logging is already covered in your PR.

MangoIV replied:
I will change it to `runServer`

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,133 @@
+delegateServerApp tick pending = do
+  connection <- liftIO $ acceptRequest pending
+  trace FrontendConnected
+  -- FIXME: we will probably have to do this in parallel, but how can we enqueue in the same
```
Yes, we discussed that before and my initial version had StateT tood. And then we dropped state so no such tricks are needed anymore.

Whole point was that we can just do dumb stateless producer threads in `IO` monad, and leave all custom monadic stuff to `delegateStep`.

Also with parallel threads we can just have one-shot thread which posts `Started` event and drop `isFirstRun` fully.

MangoIV replied:
> Whole point was that we can just do dumb stateless producer threads in IO monad, and leave all custom monadic stuff to delegateStep.

I mostly put it there because having `StateT` shows better that we pass some state that is changed later then updating it by hand. but if you want I can change it back

uhbif19 replied:
> but if you want I can change it back

We do not need that state after #218 anymore anyway.

MangoIV replied:
I removed the state as part of my changes wrt channels, we can just write a Started event to the incoming channel at the very beginning

MangoIV commented on delegate-app/HydraAuction/Delegate/Server.hs:
```diff
@@ -0,0 +1,119 @@
+
+instance Pretty DelegateServerLog where
+  pretty = \case
+    Started port -> "Started Server at Port" <> pretty (show port)
```
```suggestion
    Started port -> "Started Server at Port" <> viaShow port
```

MangoIV committed:
[fix] `viaShow` instead of `pretty . show`

MangoIV commented:
I think a possible solution for the concurrency issue is having a `TChan` that the `DelegateRunner` knows and then every `tick` read of the `TChan` all FrontendRequests, then do stuff, after that write to a `broadCastTChan` which will then be cloned by the clients as soon as they're spawned with `dupTChan`

MangoIV commented:
I will implement that tomorrow then

MangoIV commented:
Inwards it can be a TQueue, I think. Outwards I'll use TChan

nazrhom commented:
Hey @MangoIV have a look at [this pr](https://github.com/mlabs-haskell/hydra-auction/pull/218) I have already added a `Chan` to queue events in, and `delegateStep` can consume that directly.

Also - this might have been already discussed - but I don't think we should pass a tick here. Ideally we can listen for websocket connections and spawn a thread thread to handle each incoming request.

MangoIV committed:
[feat] implement concurrency
- two channels, an incoming and an outgoing one
- incoming channel receives from all clients and gets flushed by the
  consumer
- outgoing channel is a broadcast channel that gets cloned by each
  client and then sends all events back to the frontend

MangoIV committed:
[chore] refactoring and documentation
- factor out Tracing in a separate module
- remove tagless final style where sensible
- haddock

MangoIV committed:
[fix] some minor fixes
- fixes that Gregory requested
- nicer logging
- some fork off handler

MangoIV commented on hydra-auction.cabal:
```diff
@@ -183,9 +183,23 @@ executable hydra-auction
   hs-source-dirs: delegate-app
   ghc-options:    -threaded -rtsopts
-  build-depends:  hydra-auction
+  build-depends:
```
I removed these, but forgot to push

uhbif19 commented on src/HydraAuction/Types.hs:
```diff
@@ -85,11 +88,7 @@ naturalToInt (Natural i) = i
-    case intToNatural int of
-      Just nat -> return nat
-      Nothing -> fail "Integer is not natural"
+  parseJSON = maybe (fail "Integer is not natural") return . intToNatural <=< parseJSON
```
What is wrong with do notation?

MangoIV replied:
Idk, it was just something random I did because I like the fish and maybe :D

uhbif19 replied:
> because I like the fish

Yep, I did understood that :D

uhbif19 commented on delegate-app/HydraAuction/Delegate/Tracing.hs:
```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE FunctionalDependencies #-}
```
That should belong to hydra-auction-utils dir.

uhbif19 replied:
Cuz is not domain-specific.

uhbif19 commented on delegate-app/HydraAuction/Delegate/Tracing.hs:
```diff
@@ -0,0 +1,73 @@
+   type that serves as the trace and the monad
+   that you can trace in
+-}
+class MonadTracer t m | m -> t where
```
We already have one such interface. So you need integrate this one into existing one. Like make it superclass.

And I am still not understanding why fundeps are better here. Type family have one obvious advantage - it does not require to write second variable. What is advantage of fundeps?

MangoIV replied:
The obvious advantage is that you don't need to specify an equality constraint if you want a specific type to trace with.

uhbif19 replied:
Okay, you are right, that may be actual usecase if we will kill `logMsg`. And we will need ability to make message types non fixed per monad, by embedding one into another. That is also a way to kill `logMsg` - just introduce String message type instance.

uhbif19 replied:
Separated this to follow-up issue: https://github.com/mlabs-haskell/hydra-auction/issues/231

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+
+  let conf :: DelegateServerConfig
+      conf =
+        DelegateServerConfig
```
Config stuff is cool.

uhbif19 commented on delegate-app/HydraAuction/Delegate/Server.hs:
```diff
@@ -0,0 +1,78 @@
+
+-- | The config for the delegate server
+data DelegateServerConfig = DelegateServerConfig
+  { dlgt'host :: IP
```
That is style is different for all other repo. Style should be same everythere.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+runDelegateServer ::
+  -- | the configuration of the delegate server
+  DelegateServerConfig ->
+  TracerT DelegateServerLog IO ()
```
You not using your `DelegateTracerT` thing here as in other places. While I do not see why `TracerT DelegateServerLog` is worse, but still.

MangoIV replied:
I'm using it in Main

MangoIV replied:
But I can remove it if you wish.

uhbif19 replied:
Yeah, I see, I men you use it in other places but not this one => not consitent.

MangoIV replied:
Oversight. Thank you.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+  TQueue DelegateInput ->
+  -- | the broadcast queue of outgoing messages (write only)
+  TChan DelegateResponse ->
+  DelegateRunnerT (DelegateTracerT IO) ()
```
I think that term `Runner` is already used as name for monad pattern we use, not for actual code block.

MangoIV replied:
I think I don't understand this.

uhbif19 replied:
We use word `Runner` to name monad pattern. You name kind of thread the same term. But it is different meaning for same term - we use `Runner` monad(s) in multiple different situations. That is such term selection is confusing.

uhbif19 reviewed (changes requested):
Still think `mbQueueAuctionPhases` is not best way to separate function.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+  -- | the broadcast queue of outgoing messages (write only)
+  TChan DelegateResponse ->
+  DelegateRunnerT (DelegateTracerT IO) ()
+-- FIXME: we need to abort at some point but this doesn't seem
```
Also we do not handle Ctrl-C reaction and thread finalization stuff at all. I am not aware what is proper way to deal with Ctrl-C.

MangoIV replied:
I'm not 100% confident but I think this is resolved now.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+   subscribes to the broadcast channel and sends to
+   the incoming channel of the delegate runner
+-}
+delegateServerApp ::
```
I do not understand naming of this function.

MangoIV replied:
It's a ServerApp and specifically the serverApp that is run by the delegate.

uhbif19 replied:
What is app in it?

MangoIV replied:
well it's a ServerApp, which is a function from `PendingConnection -> IO ()` https://hackage.haskell.org/package/websockets-0.12.7.3/docs/Network-WebSockets.html#t:ServerApp

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+
+  _ <-
+    liftIO . forkIO $
+      runServer (show $ dlgt'host conf) (fromIntegral $ dlgt'port conf) $
```
> fromIntegral

Why not just store it in that type already in config?

MangoIV replied:
Because it's basically untyped. The proper server I had before took proper types. This doesn't. If you wish I can change it. Be aware that we don't validate IPs then anymore (which we do now by parsing IPs)

uhbif19 replied:
Ah, there is some specialized type in config, understand now, you right.

uhbif19 replied:
Ah, there is some specific port type in config, understand now, you right.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+      runServer (show $ dlgt'host conf) (fromIntegral $ dlgt'port conf) $
+        runWithTracer' tracer . delegateServerApp (dlgt'ping conf) delegateInputChan delegateBroadCastChan
+
+  execDelegateRunnerT $ mkRunner (dlgt'tick conf) delegateInputChan delegateBroadCastChan
```
This still looks like it only works once.

MangoIV replied:
What do you mean it works only once? This replies to requests forever.

uhbif19 replied:
I was wrong, missed forever.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+    let encodeSend :: forall a. ToJSON a => a -> IO ()
+        encodeSend = sendTextData connection . encode
+
+        receive = forever $
```
Please move this in `where` blocks or separate functions.

MangoIV replied:
These use parameters that are brought in scope in the do block. These are not in scope in a where. They would have to be functions that take two parameters which imho is unnecessarily verbose.

uhbif19 replied:
Yes. I know. This makes dependencies explicit which is not that bad. Also log do is bad. I tried to explain my position on this here: https://app.slack.com/client/T01UN827FQV/C04C3ASUFEF/thread/C04C3ASUFEF-1680260404.710759

We can leave `let` stuff now and reach consensus on that late.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+            case eitherDecode @FrontendRequest inp of
+              Left msg -> do
+                let err = FrontendNoParse msg
+                trace (DelegateError err)
```
I do not think that this should be named `DelegateError`. This is error on client side, for delegate it is just normal event.

MangoIV replied:
The delegate runner t doesn't see this. This follows the pattern of naming a constr the same as the type that it takes that you use in the project. Err is of type DelegateError.

uhbif19 replied:
Yes, you can just move that consturctor- from `DelegateError` type.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+        lift $ traverse_ (trace . DelegateOutput) responses
+
+  events <- liftIO . atomically $ flushTQueue reqq
+  traverse_ (delegateStep >=> encodeBroadcast) events
```
Your PR does not address broadcast/per-client output distinction problem we discussed earlier.

MangoIV replied:
I wasn't aware that was in scope.

MangoIV replied:
this is handled by a follow up PR

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+        liftIO . atomically $ traverse_ (writeTChan broadcast) responses
+        lift $ traverse_ (trace . DelegateOutput) responses
+
+  events <- liftIO . atomically $ flushTQueue reqq
```
I guess you right that there are the cases when this works better, but this does not "solve" queue overload issue. I think that PR is right place to document that in `off_chain_spec.md`. Or I can do it in my PR.

MangoIV replied:
Yes. It only solves it if we can assume that the tick is the time consuming part.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+-- FIXME: we need to abort at some point but this doesn't seem
+-- to be implemented yet so we just go on
+mkRunner tick reqq broadcast = forever $ do
+  let encodeBroadcast responses = do
```
Please use `where`.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+  DelegateRunnerT (DelegateTracerT IO) ()
+-- FIXME: we need to abort at some point but this doesn't seem
+-- to be implemented yet so we just go on
+mkRunner tick reqq broadcast = forever $ do
```
```suggestion
mkRunner tick delegateInputChan broadcast = forever $ do
```

uhbif19 replied:
Also, why Chan if it is queue.

MangoIV replied:
A TChan is just a Queue with a different implementation. I named it reqq for request queue.

MangoIV replied:
I think semantically the names are arbitrary and interchangeable. I don't know why the name "channel" would imply that it is implemented as an actual distributed list whereas the Queue is implemented as a single TVar holding a normal list

uhbif19 replied:
Okay.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+  liftIO $ do
+    broadcast' <- atomically $ dupTChan broadcast
+
+    let encodeSend :: forall a. ToJSON a => a -> IO ()
```
```suggestion
    let encodeAndSend :: forall a. ToJSON a => a -> IO ()
```

MangoIV replied:
I don't understand how this would improve the function or add additional semantics to the name.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -1,49 +1,179 @@
+            atomically (readTChan broadcast')
+              >>= encodeSend
+
+    withPingThread connection pingSecs (pure ()) $ race_ receive send -- TODO: can we do this
```
> -- Many (but not all) proxies have a 60 second default timeout, so based on that
> -- sending a ping every 30 seconds is a good idea.

Please add code comment on why code uses this ping stuff.

MangoIV committed:
Merge remote-tracking branch 'origin/staging' into mangoiv/175-implement-real-network-interface-to-be-used-by-frontend-for-delegate-server

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -152,12 +165,55 @@ runDelegateServer conf = do
+
+mbQueueAuctionPhases :: TQueue DelegateInput -> TChan DelegateResponse -> DelegateTracerT IO ()
+mbQueueAuctionPhases reqq =
+  liftIO . atomically . (dupTChan >=> getTerms)
```
Please write such things in `do`-notation.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -152,12 +165,55 @@ runDelegateServer conf = do
+        AuctionSet terms -> pure terms
+        _ -> getTerms chan
+
+    queueCurrentStage :: AuctionTerms -> IO ()
```
Name is misleading, this is forever-running thread.

uhbif19 replied:
This should be somehow marked by name. `queueCurrentStage` sounds like one-action function.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -152,12 +165,55 @@ runDelegateServer conf = do
-        runWithTracer' tracer . delegateServerApp (dlgt'ping conf) delegateInputChan delegateBroadCastChan
-
-  execDelegateRunnerT $ mkRunner (dlgt'tick conf) delegateInputChan delegateBroadCastChan
+  let startWorker :: ThreadSort -> IO () -> IO ()
```
I think this should be standalone function.

Defunctionalization is fun. But I guess this wont cover the difference of queues which will follow input queue will be separated.

uhbif19 commented on delegate-app/Main.hs:
```diff
@@ -152,12 +165,55 @@ runDelegateServer conf = do
+mbQueueAuctionPhases reqq =
+  liftIO . atomically . (dupTChan >=> getTerms)
+    >=> liftIO . (traceQueueEvent *> queueCurrentStage)
+  where
```
Thanks.

uhbif19 commented on delegate-app/HydraAuction/Delegate/Server.hs:
```diff
@@ -46,6 +49,9 @@ data DelegateServerLog
   | FrontendInput FrontendRequest
   | DelegateOutput DelegateResponse
   | DelegateError DelegateError
+  | StartThread ThreadSort
```
Start/Cancel could be merged I guess.

MangoIV replied:
How would we merge these?

uhbif19 replied:
Like `ThreadEvent Kind ThreadSort`

MangoIV committed:
[fix] resolve some remarks by Gregory

MangoIV committed:
Merge remote-tracking branch 'origin/staging' into mangoiv/175-implement-real-network-interface-to-be-used-by-frontend-for-delegate-server

MangoIV committed:
[fix] move Tracing to HydraAuctionUtils

MangoIV committed:
[fix] some fixes as asked for by Gregory and some streamlining
- derivive pretty via show instead of using viaShow directly
- more documentation
- more where clauses where possible
- separation of the startWorker function
- factor out some part of the serverlog

nazrhom commented on delegate-app/Main.hs:
```diff
@@ -1,62 +1,254 @@
-          threadDelay (fromInteger s * 1000)
-          queueCurrentStage
-mbQueueAuctionPhases _ _ = pure ()
+          threadDelay (fromInteger s * tick)
```
I don't think this is right... The 1000 is meant to convert seconds to milliseconds, so threadDelay waits the correct amount of time

MangoIV replied:
Oh. Thank you. Will fix

nazrhom commented:
Looks like all conversations are resolved here. LGTM
